### PR TITLE
#100 - Fix banner dismissal issues

### PR DESCRIPTION
**Pull request by mattbarker016**
#100. This resolves an issue where banners for connectively wouldn't properly dismiss after re-connecting to the internet. I also refactored the code around banner showing, dismissal, and reachability changes, resulting in less variables and a more straight forward approach.

### CODE REVIEW

File: TCAT/Controllers/AllStopsTableViewController.swift [hunk 1]
```diff
@@ -186,4 +205,32 @@ extension AllStopsTableViewController: DZNEmptyDataSetSource {
...
     }
+    
+    func buttonTitle(forEmptyDataSet scrollView: UIScrollView!, for state: UIControl.State) -> NSAttributedString! {
+        let title = "Retry"
+        let attrs = [NSAttributedString.Key.foregroundColor: UIColor.buttonColor]
+        return NSAttributedString(string: title, attributes: attrs)
+    }
+    
+    func emptyDataSet(_ scrollView: UIScrollView!, didTapButton: UIButton!) {
+        print("button tapped")
```
Comment (AAAstorga) [inline TCAT/Controllers/AllStopsTableViewController.swift]: remove print please

File: TCAT/Controllers/AllStopsTableViewController.swift [hunk 2]
```diff
@@ -107,11 +105,32 @@ class AllStopsTableViewController: UITableViewController {
...
+            // Adding "#" to keys for bus stops that start with a number
+            sortedKeys.append("#")
+        }
+        
+        return sortedKeys
+    }
+    
+    func setUpTableOnRetry() {
+        // Retry getting data from user defaults
+        print("retry getting data from user defaults")
```
Comment (AAAstorga) [inline TCAT/Controllers/AllStopsTableViewController.swift]: remove the print statement

File: TCAT/Controllers/AllStopsTableViewController.swift [hunk 3]
```diff
@@ -186,4 +205,32 @@ extension AllStopsTableViewController: DZNEmptyDataSetSource {
...
+    
+    func buttonTitle(forEmptyDataSet scrollView: UIScrollView!, for state: UIControl.State) -> NSAttributedString! {
+        let title = "Retry"
+        let attrs = [NSAttributedString.Key.foregroundColor: UIColor.buttonColor]
+        return NSAttributedString(string: title, attributes: attrs)
+    }
+    
+    func emptyDataSet(_ scrollView: UIScrollView!, didTapButton: UIButton!) {
+        print("button tapped")
+        retryNetwork{() -> Void in
```
Comment (AAAstorga) [inline TCAT/Controllers/AllStopsTableViewController.swift]: space after `retryNetwork {`

File: TCAT/Controllers/AllStopsTableViewController.swift [hunk 4]
```diff
@@ -186,4 +205,32 @@ extension AllStopsTableViewController: DZNEmptyDataSetSource {
...
+            self.setUpTableOnRetry()
+        }
+    }
+    
+    func retryNetwork(completion: @escaping () -> Void) {
+        print("retry network")
```
Comment (AAAstorga) [inline TCAT/Controllers/AllStopsTableViewController.swift]: remove print

File: TCAT/Controllers/HomeViewController.swift [hunk 5]
```diff
@@ -129,29 +127,26 @@ class HomeViewController: UIViewController {
...
             switch reachability.connection {
             case .none:
-                if !isBannerShown {
-                    banner = StatusBarNotificationBanner(title: Constants.Banner.noInternetConnection, style: .danger)
-                    banner!.autoDismiss = false
-                    banner!.show(queuePosition: .front, on: navigationController)
-                    isBannerShown = true
-                }
+                banner = StatusBarNotificationBanner(title: Constants.Banner.noInternetConnection, style: .danger)
+                banner!.autoDismiss = false
```
Comment (AAAstorga) [inline TCAT/Controllers/HomeViewController.swift]: I know the banner exists, but can we please do an `if let banner = banner {...}` Would appreciate it!

File: TCAT/Controllers/RouteDetail+ContentViewController.swift [hunk 6]
```diff
@@ -227,24 +227,22 @@ class RouteDetailContentViewController: UIViewController, GMSMapViewDelegate, CL
...
-        if self.banner == nil {
-            self.banner = StatusBarNotificationBanner(title: message, style: status)
-            self.banner!.autoDismiss = false
-            self.banner!.dismissOnTap = true
-            self.banner!.show(queuePosition: .front, on: navigationController)
-            self.isBannerShown = true
-        }
+        hideBanner()
+        banner = StatusBarNotificationBanner(title: message, style: status)
+        banner!.autoDismiss = false
```
Comment (AAAstorga) [inline TCAT/Controllers/RouteDetail+ContentViewController.swift]: same if let here as well please

File: TCAT/Controllers/RouteOptionsViewController.swift [hunk 7]
```diff
@@ -74,13 +74,12 @@ class RouteOptionsViewController: UIViewController, UITableViewDelegate, UITable
...
-    var banner: StatusBarNotificationBanner? = nil
-
-    var isBannerShown = false {
+    var banner: StatusBarNotificationBanner? {
         didSet {
             setNeedsStatusBarAppearanceUpdate()
         }
     }
+
     var cellUserInteraction: Bool = true
```
Comment (AAAstorga) [inline TCAT/Controllers/RouteOptionsViewController.swift]: don't need `: Bool`

File: TCAT/Controllers/RouteOptionsViewController.swift [hunk 8]
```diff
@@ -152,23 +151,19 @@ class RouteOptionsViewController: UIViewController, UITableViewDelegate, UITable
...
+        reachability?.stopNotifier()
+        NotificationCenter.default.removeObserver(self, name: .reachabilityChanged, object: reachability)
+        // Remove banner
+        banner?.dismiss()
+        banner = nil
         // Deactivate and remove timers
-        for (_, timer) in timers {
-            timer.invalidate()
-        }
+        for (_, timer) in timers { timer.invalidate() }
```
Comment (AAAstorga) [inline TCAT/Controllers/RouteOptionsViewController.swift]: timers.values.forEach { $0.invalidate() }

File: TCAT/Utilities/SearchTableViewHelpers.swift [hunk 9]
```diff
@@ -58,15 +58,18 @@ class SearchTableViewManager {
 
     private init() {}
     func getAllStops() -> [BusStop] {
-        if let stops = allStops {
-            return stops
+        if let stops = allStops{
```
Comment (AAAstorga) [inline TCAT/Utilities/SearchTableViewHelpers.swift]: space after `allStops {` Make sure to run `swiftlint autocorrect` before finishing this PR

**Review by AAAstorga — APPROVED** [review]
Looks good! Just had a few stylistic issues. Make sure everyone is running `swiftlint autocorrect` before ideally making a PR. It would of fixed some of the styling errors

File: TCAT/Controllers/RouteOptionsViewController.swift (hunk 8 above)
Comment (mattbarker016) [reply TCAT/Controllers/RouteOptionsViewController.swift]: Ah I tried to do something like that, values was what I'm looking for, thanks!

**Comment by mattbarker016** [comment]
@AAAstorga I actually branched this off of Yana's work, but we can nab these fixes now. I should have set up the base to be her branch, but we can tackle the feedback.

File: TCAT/Controllers/RouteOptionsViewController.swift (hunk 8 above)
Comment (mattbarker016) [reply TCAT/Controllers/RouteOptionsViewController.swift]: Resolved

File: TCAT/Controllers/RouteOptionsViewController.swift (hunk 7 above)
Comment (mattbarker016) [reply TCAT/Controllers/RouteOptionsViewController.swift]: Resolved, plus other little related stuff

File: TCAT/Controllers/HomeViewController.swift (hunk 5 above)
Comment (mattbarker016) [reply TCAT/Controllers/HomeViewController.swift]: @AAAstorga I reset the banner to be nil in this function to reset it properly. Would it be better to just use `?` for lines 143 and such? `if let` would be pretty redundant, I'd have to set the variable and then check if it's not nil.

**Comment by mattbarker016** [comment]
@ywsang Make sure you grab anything I didn't resolve. Austin mentioned running `swiftlint autocorrect`, but I would run that in Terminal **for the specific files you touched**, otherwise there will be a lot to fix. And that's what #139 is all about 😃

**Comment by mattbarker016** [comment]
Actually, I'm going to just merge my changes into your branch @ywsang, but make sure you make note of Austin's suggestions once you finish your feature and aren't debugging!

**Comment by mattbarker016** [comment]
I'll leave this open until Austin's suggestions are taken care of!

**Pull request merged**